smarts/core/traffic_history.py

```diff
@@ -230,6 +297,127 @@ def vehicles_active_between(
+                    FROM (SELECT vehicle_id, sim_time, speed from Trajectory WHERE sim_time >= ? AND sim_time < ?)
+                    GROUP BY vehicle_id
+                   ) AS D ON V.id = D.vehicle_id
+                   JOIN (
+                    {self._greatest_n_per_group_format(
+                        select='''vehicle_id,
+                        sim_time,
+                        speed,
+                        position_x,
+                        position_y,
+                        heading_rad''',
+                        table='Trajectory',
+                        group_by='vehicle_id',
+                        greatest_of_group='sim_time',
+                        where='sim_time >= ?',
+                        operation="MIN"
+                    )}
+                   ) AS S ON V.id = S.vehicle_id
+                   JOIN (
+                    {self._greatest_n_per_group_format(
+                        select='''vehicle_id,
+                        sim_time,
+                        speed,
+                        position_x,
+                        position_y,
+                        heading_rad''',
+                        table='Trajectory',
+                        group_by='vehicle_id',
+                        greatest_of_group='sim_time',
+                        where='sim_time < ?'
+                    )}
+                   ) AS E ON V.id = E.vehicle_id
+                   WHERE E.sim_time - S.sim_time >= ?
+                   GROUP BY V.id
+                   ORDER BY S.sim_time
+                   """
+
+        rows = self._query_list(
+            query,
+            (
+                exists_at_or_after,
+                ends_before,
+                exists_at_or_after,
+                ends_before,
+                minimum_vehicle_window,
+            ),
+        )
+
+        def _slice_from_row(row):
+            return TrafficHistory.TrafficHistoryVehicleWindow(
+                row[0],
+                self.decode_vehicle_type(row[1]),
+                *self._resolve_vehicle_dims(row[1], *row[2:5]).as_lwh,
+                *row[5:],
+            )
+
+        def debug(rows):
+            seen = set()
+            seen_pos_x = set()
+            rs = list()
+            for row in rows:
+                r = _slice_from_row(row)
+                assert r.vehicle_id not in seen
+                assert r.end_time - r.start_time >= minimum_vehicle_window
+                assert r.start_time >= exists_at_or_after
+                assert r.end_time <= ends_before
+                assert r.end_position_x not in seen_pos_x
+                seen_pos_x.add(r.end_position_x)
+                seen.add(r.vehicle_id)
+                rs.append(r)
+            self._log.info(f"Num vehicles suiting time criteria: {len(rs)}")
+            return rs
+
+        thvtss = None
+        thvtss = debug(rows)
+
+        return thvtss or (_slice_from_row(row) for row in rows)
```

Comment:
Is this just to allow easy toggling on/off of the debugging path? It's a bit confusing because one is a list and one is a generator.

Reply:
Yes, I think it is badly addressed, I will swap to including the debugging in the generator instead of resolving it to a list.